util/time.go

```diff
@@ -0,0 +1,42 @@
+// Copyright 2018 Google Inc. All Rights Reserved.
+//
+// Licensed under the Apache License, Version 2.0 (the "License");
+// you may not use this file except in compliance with the License.
+// You may obtain a copy of the License at
+//
+//     http://www.apache.org/licenses/LICENSE-2.0
+//
+// Unless required by applicable law or agreed to in writing, software
+// distributed under the License is distributed on an "AS IS" BASIS,
+// WITHOUT WARRANTIES OR CONDITIONS OF ANY KIND, either express or implied.
+// See the License for the specific language governing permissions and
+// limitations under the License.
+
+package util
+
+import (
+	"context"
+	"time"
+)
+
+// Sleep sleeps for at least the specified duration. Returns false iff done is
+// closed before the timer fires.
+func Sleep(done <-chan struct{}, dur time.Duration) bool {
+	timer := time.NewTimer(dur)
+	select {
+	case <-timer.C:
+		return true
+	case <-done:
+		timer.Stop()
```

Comment:
Out of interest, why does the timer get stopped in this case, but not in the other case?  Is the timer going to continue to be used in the other case?  Doesn't look like it can be as it's local to this function...

Reply:
`Stop`, as it says in its doc, prevents the timer from firing. In the `<-timer.C` case it has already fired, so no need to prevent it. However, I might be wrong in thinking that not calling `Stop` is okay in terms of cleaning up timer resources. Couldn't quickly find info on how to do that properly.

Reply:
The docs say you need to call stop to release resources. Is there a leak here?

Reply:
Alternate implementation idea: use `sleepCtx := context.WithTimeout(ctx, dur)`, and just block on `sleep.Ctx.Done()`? Then you can just return `ctx.Err()`, it will be nil if it was just the `sleepCtx` that finished, which is exactly what you want.